cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java

```diff
@@ -158,27 +161,72 @@ public FileUploadHandler(File outputDir) {
     public HttpContentConsumer upload(HttpServiceRequest request,
                                       HttpServiceResponder responder,
                                       @PathParam("file") String file) throws IOException {
-      final WritableByteChannel channel = Channels.newChannel(new FileOutputStream(new File(outputDir, file)));
-      return new HttpContentConsumer() {
-        @Override
-        public void onReceived(ByteBuffer chunk, Transactional transactional) throws Exception {
-          channel.write(chunk);
-        }
+      return new FileContentConsumer(new File(outputDir, file));
+    }
 
-        @Override
-        public void onFinish(HttpServiceResponder responder) throws Exception {
-          channel.close();
-          responder.sendStatus(200);
-        }
+    @Path("/download/{file}")
+    @GET
+    public void download(HttpServiceRequest request,
+                         HttpServiceResponder responder,
+                         @PathParam("file") String file) {
+      Location location = Locations.toLocation(new File(outputDir, file));
+      try {
+        responder.send(200, location, "text/plain");
+      } catch (IOException e) {
+        responder.sendStatus(500);
+      }
+    }
 
+    // A POST endpoint for upload file and response with the file content using HttpContentProducer
+    @Path("/upload/{file}")
+    @POST
+    public HttpContentConsumer uploadDownload(HttpServiceRequest request,
+                                              HttpServiceResponder responder,
+                                              @PathParam("file") String file) throws IOException {
+      final File targetFile = new File(outputDir, file);
+      return new FileContentConsumer(targetFile) {
         @Override
-        public void onError(HttpServiceResponder responder, Throwable failureCause) {
-          LOG.error("Failed when handling upload", failureCause);
+        protected void response(HttpServiceResponder responder) throws IOException {
```

Comment:
this first reads the entire request body, then sends back the response. Is it actually possible to interleave the receiving and sending? For example, you read a request that has a huge list of keys in its body; for each key you look it up in some table and send key:value back. In that case, you would want to start sending at the same as you are still receiving